### PR TITLE
Allow to read parquet binary column as UTF8 type

### PR DESCRIPTION
**Pull request by goldmedal**
# Which issue does this PR close?

No related issue.
# Rationale for this change
While working on https://github.com/apache/datafusion/issues/12788#issuecomment-2402885802 in DataFusion, I found we can't read the parquet binary column as string types (Utf8, LargeUtf8, or Utf8View) through `ArrowReaderOptions::with_schema`. I think it makes sense to read them as strings if the user ensures it's a string binary value.

# What changes are included in this PR?
I added some matching rules in `apply_hint` in `parquet/src/arrow/schema/primitive.rs` to handle the binary-to-string cases.

# Are there any user-facing changes?
no

cc @alamb

### CODE REVIEW

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 1]
```diff
@@ -3077,6 +3077,119 @@ mod tests {
...
+        let options = ArrowReaderOptions::new().with_schema(Arc::new(Schema::new(supplied_fields)));
+        let mut arrow_reader = ParquetRecordBatchReaderBuilder::try_new_with_options(
+            file.try_clone().unwrap(),
+            options,
+        )
+        .expect("reader builder with schema")
+        .build()
+        .expect("reader with schema");
+
+        arrow_reader.next();
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:3190]:
```suggestion
        arrow_reader.next().unwrap_err();
```

As this should error, given the data isn't actually UTF-8

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 2]
```diff
@@ -3077,6 +3077,119 @@ mod tests {
...
+        let batch = arrow_reader.next().unwrap().unwrap();
+        assert_eq!(batch.num_columns(), 3);
+        assert_eq!(batch.num_rows(), 3);
+        assert_eq!(
+            batch
+                .column(0)
+                .as_any()
+                .downcast_ref::<StringArray>()
+                .expect("downcast to string")
+                .iter()
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:3136]:
```suggestion
                .column(0)
                .as_string::<i32>()
                .iter()
```

And the same below

File: parquet/src/arrow/schema/primitive.rs [hunk 3]
```diff
@@ -57,6 +57,11 @@ fn apply_hint(parquet: DataType, hint: DataType) -> DataType {
         (DataType::Utf8, DataType::LargeUtf8) => hint,
         (DataType::Binary, DataType::LargeBinary) => hint,
 
+        // Read as Utf8
```
Comment (alamb) [inline parquet/src/arrow/schema/primitive.rs:60]: ❤️

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 4]
```diff
@@ -3077,6 +3077,108 @@ mod tests {
...
+        );
+
+        assert_eq!(
+            batch.column(2).as_string_view().iter().collect::<Vec<_>>(),
+            vec![Some("one"), Some("two"), Some("three")]
+        );
+    }
+
+    #[test]
+    #[should_panic(expected = "Invalid UTF8 sequence at")]
```
Comment (alamb) [inline parquet/src/arrow/arrow_reader/mod.rs:3155]: ❤️

**Review by alamb — APPROVED** [review]
Thank you @goldmedal and @tustvold  -- this looks great to me

**Comment by goldmedal** [comment]
Thanks @alamb  @tustvold !

**Pull request merged**